File: 2019/python/day03.py
def get_wire_coords(wire):
    coords = [0, 0]

    for point in wire.split(','):
        for _ in range(int(point[1:])):
            idx = 0 if point[0] in ('L', 'R') else 1
            coords[idx] += -1 if point[0] in ('L', 'D') else 1

            yield tuple(coords)


with open('inputs/day03.txt') as input:
    wires = [list(get_wire_coords(wire))
             for wire in (line.strip() for line in input)]

intersections = set(wires[0]).intersection(set(wires[1]))

part_one = min((abs(x) + abs(y) for (x, y) in intersections))

part_two = min(sum(wire.index(intersection) + 1 for wire in wires)
               for intersection in intersections)

print(part_one)  # 1285

print(part_two)  # 14228

File: 2019/javascript/day03.js
const readInput = require('./read-input')

const input = readInput('day03.txt')

const processWire = (wire) => {
  const coords = [0, 0]
  const allSteps = new Set()

  for (const point of wire.split(',')) {
    for (_ = 0; _ < Number(point.slice(1)); _++) {
      const idx = ['L', 'R'].includes(point.charAt(0)) ? 0 : 1
      coords[idx] += ['L', 'D'].includes(point.charAt(0)) ? -1 : 1
      allSteps.add(`${coords[0]},${coords[1]}`)
    }
  }

  return allSteps
}

const wires = input.split('\n').map(processWire)
const intersections = []
wires[0].forEach((x) => wires[1].has(x) && intersections.push(x))

const partOne = Math.min(...intersections.map((i) => Math.abs(Number(i.split(',')[0])) + Math.abs(Number(i.split(',')[1]))))
const partTwo = Math.min(...intersections.map((i) => wires.reduce((s, w) => s + [...w].indexOf(i) + 2, 0)))

console.log('partOne:', partOne) // 1285
console.log('partTwo:', partTwo) // 14228
